README
List-Unique-DeterministicOrder

This module provides a structure to store a set
of keys, without duplicates, and be able to access
them by either key name or index.



Changes
Revision history for List-Unique-DeterministicOrder

0.004    2023-10-05
    - Micro-optimise the deletion methods

0.003    2022-09-24
    - Fix class returned from clone 

0.002    2022-09-24
    - Provide our own clone method

0.001    2018-08-17
    - First implementation.



Makefile.PL
use 5.010;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'List::Unique::DeterministicOrder',
    AUTHOR           => q{SLAFFAN <shawnlaffan@gmail.com>},
    VERSION_FROM     => 'lib/List/Unique/DeterministicOrder.pm',
    ABSTRACT_FROM    => 'lib/List/Unique/DeterministicOrder.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.010',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More' => '0',
        'rlib' => 0,
        'List::Util' => 0,
    },
    PREREQ_PM => {
        'List::Util' => 1.45,
        'Carp' => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/shawnlaffan/perl-List-Unique-DeterministicOrder',
            },
            bugtracker => {
                web => 'https://github.com/shawnlaffan/perl-List-Unique-DeterministicOrder/issues/',
            },
        },
    },

    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'List-Unique-DeterministicOrder-*' },
);


etc/bench/bench.pl

use Benchmark qw {:all};
use 5.016;
use Data::Dumper;
use Test::More;

use rlib '../../lib';

use List::Unique::DeterministicOrder;
use List::BinarySearch::XS qw /binsearch/;
use List::MoreUtils::XS 0.423 qw/bremove binsert/;
use List::MoreUtils;# qw /bremove/;
use Hash::Ordered;

use Clone qw /clone/;


my $nreps = $ARGV[0] || 500;
my $data_size = $ARGV[1] || 1000;
my $run_benchmarks = $ARGV[2];

#  ratio of insertions to deletions
my $insertion_frac = 0.1;
my $insert_count   = $data_size * $insertion_frac;

srand 1534390472;

my %hashbase;
#@hash{1..1000} = (rand()) x 1000;
my $item = 'a';
for my $i (1 .. ($data_size +$insert_count)) {
    $hashbase{$item} = rand() + 1;
    $item ++;
}
my $hashref = \%hashbase;

my @rand_keys    = sort {$hashbase{$a} <=> $hashbase{$b}} keys %hashbase;
my @insertions   = splice @rand_keys, 0, $data_size * $insertion_frac;
my @sorted_keys  = sort @rand_keys;
my @sorted_pairs = map {$_ => 1} @sorted_keys;



my $dds_base = List::Unique::DeterministicOrder->new(data => \@sorted_keys);
my $ho_base  = Hash::Ordered->new (@sorted_pairs);

my %data = (
    lmu => [],
    lbs => [],
    ldd => [],
    lho => [],
    baseline => [],
);

#  make lots of copies to ensure data generation
#  is outside the benchmarking
foreach my $i (0 .. $nreps+1) {
    push @{$data{lmu}}, [@sorted_keys];
    push @{$data{lbs}}, [@sorted_keys];
    push @{$data{ldd}}, clone $dds_base;
    push @{$data{lho}}, clone $ho_base;
    push @{$data{baseline}}, [@sorted_keys];
}




my $l1 = lmu();
my $l2 = lbs();
my $l3 = ldd();
my $l4 = lho();

say 'First few items in each list:';
say join ' ', @$l1[0 .. 5];
say join ' ', @$l2[0 .. 5];
say join ' ', @$l3[0 .. 5];
say join ' ', @$l4[0 .. 5];

is_deeply ($l1, $l2, 'same order');
is_deeply ($l1, [sort @$l3], 'same contents, list-u-det-order');
is_deeply ($l1, [sort @$l4], 'same contents, hash ordered');


done_testing();

exit if !$run_benchmarks;


cmpthese (
    $nreps,
    {
        lmu  => sub {lmu()},
        lbs  => sub {lbs()},
        ldd  => sub {ldd()},
        lho  => sub {lho()},
        baseline => sub {baseline()},
    }
);

sub lbs {
    my $list = shift @{$data{lbs}};
    my $i = -1;
    foreach my $key (@rand_keys) {
        $i++;
        delete_from_sorted_list_aa($key, $list);
        my $insert = $insertions[$i] // next;
        binsert {$_ cmp $insert} $insert, @$list;
    }
    $list;
}

sub delete_from_sorted_list_aa {
    my $idx  = binsearch { $a cmp $b } $_[0], @{$_[1]};
    splice @{$_[1]}, $idx, 1;

    $idx;
}

sub insert_into_sorted_list_aa {
    my $idx  = binsearch_pos { $a cmp $b } $_[1], @{$_[2]};
    splice @{$_[2]}, $idx, 0, $_[1];

    $idx;
}


sub lmu {
    my $list = shift @{$data{lmu}};
    my $i = -1;
    foreach my $key (@rand_keys) {
        $i++;
        bremove {$_ cmp $key} @$list;
        my $insert = $insertions[$i] // next;
        binsert {$_ cmp $insert} $insert, @$list;
    }
    $list;
}

sub ldd {
    #  $dds reflects the old name for the module
    my $dds = shift @{$data{ldd}};
    my $i = -1;
    foreach my $key (@rand_keys) {
        $i++;
        $dds->delete ($key);
        my $insert = $insertions[$i] // next;
        $dds->push ($insert);
    }
    [$dds->keys];
}

sub lho {
    my $ho = shift @{$data{lho}};
    my $i = -1;
    foreach my $key (@rand_keys) {
        $i++;
        $ho->delete ($key);
        my $insert = $insertions[$i] // next;
        $ho->set ($insert => 1);
    }
    [$ho->keys];
}

sub baseline {
    my $list = shift @{$data{baseline}};
    my $i;
    foreach my $key (@rand_keys) {
        $i++;
        my $insert = $insertions[$i] // next;
    }
    $list;
}

__END__

This first test fluctuates a fair bit,
but List::Unique::DeterministicOrder is consistently fastest by ~40% or more.
The others have not been tested with differing PRNG seeds.  

perl etc\bench\bench.pl 5000 1000 1
First few items in each list:
aak aap aas abc abk aby
aak aap aas abc abk aby
fs agm ahh aft tp px
gz aak amq nn cp sb
ok 1 - same order
ok 2 - same contents, list-u-det-order
ok 3 - same contents, hash ordered
1..3
           Rate      lho      lbs      lmu      ldd baseline
lho       284/s       --     -11%     -11%     -73%     -96%
lbs       318/s      12%       --      -0%     -70%     -96%
lmu       319/s      12%       0%       --     -70%     -96%
ldd      1053/s     271%     231%     230%       --     -85%
baseline 7112/s    2405%    2134%    2132%     576%       --


perl etc\bench\bench.pl 500 10000 1
First few items in each list:
aak aap aas abc abk aby
aak aap aas abc abk aby
pau bpx ism hpl blm ofb
ovt loc eug gxs gz biw
ok 1 - same order
ok 2 - same contents, list-u-det-order
ok 3 - same contents, hash ordered
1..3
           Rate      lmu      lbs      lho      ldd baseline
lmu      6.60/s       --     -30%     -62%     -81%     -98%
lbs      9.36/s      42%       --     -45%     -73%     -97%
lho      17.1/s     160%      83%       --     -51%     -95%
ldd      35.1/s     431%     274%     105%       --     -90%
baseline  368/s    5479%    3829%    2047%     950%       --


perl etc\bench\bench.pl 50 50000 1
First few items in each list:
aaan aabb aabq aabx aacz aadd
aaan aabb aabq aabx aacz aadd
ahcx ansq bsir bkss aadw apiq
bmzy ovt anya bclp aijn bqtc
ok 1 - same order
ok 2 - same contents, list-u-det-order
ok 3 - same contents, hash ordered
1..3
            Rate      lmu      lbs      lho      ldd baseline
lmu      0.353/s       --     -18%     -89%     -97%    -100%
lbs      0.428/s      21%       --     -87%     -96%    -100%
lho       3.33/s     843%     678%       --     -71%     -97%
ldd       11.6/s    3168%    2595%     247%       --     -89%
baseline   107/s   30067%   24780%    3098%     823%       --


perl etc\bench\bench.pl 5 100000 1
First few items in each list:
aaan aabb aabq aabx aacz aadd
aaan aabb aabq aabx aacz aadd
cxwa aqvg kpv bfru eiwd nbe
bmzy ovt anya bclp aijn dmmr
ok 1 - same order
ok 2 - same contents, list-u-det-order
ok 3 - same contents, hash ordered
1..3
            (warning: too few iterations for a reliable count)
                Rate      lbs      lmu      lho      ldd baseline
lbs      9.73e-002/s       --      -8%     -94%     -97%    -100%
lmu          0.105/s       8%       --     -94%     -97%    -100%
lho           1.62/s    1570%    1444%       --     -53%     -96%
ldd           3.48/s    3476%    3207%     114%       --     -92%
baseline      45.9/s   47048%   43492%    2724%    1218%       --


etc/bench/profile.pl

use Benchmark qw {:all};
use 5.016;
use Data::Dumper;
use Test::More;

use rlib '../../lib';

use List::Unique::DeterministicOrder;
use List::BinarySearch::XS qw /binsearch/;
use List::MoreUtils::XS 0.423 qw/bremove binsert/;
use List::MoreUtils;# qw /bremove/;

use Clone qw /clone/;


my $nreps = $ARGV[0] || 500;
my $data_size = $ARGV[1] || 1000;


srand 1534390472;

my %hashbase;
#@hash{1..1000} = (rand()) x 1000;
for my $i (1..$data_size) {
    $hashbase{$i} = rand() + 1;
}
my $hashref = \%hashbase;

my %insertion_hash;
for my $i ('a' .. 'zzzz') {
    $insertion_hash{$i} = rand() + 1;
}
my @insertions = sort {$insertion_hash{$a} <=> $insertion_hash{$b}} keys %insertion_hash;


my @sorted_keys = sort keys %$hashref;

my $dds_base = List::Unique::DeterministicOrder->new(data => \@sorted_keys);

my %data = (
    ldd => [],
);

#  make lots of copies to ensure data generation
#  is outside the benchmarking
foreach my $i (0 .. $nreps+1) {
    push @{$data{ldd}}, clone $dds_base;
}


for my $iter (0 .. $nreps) {
    ldd();
}


sub ldd {
    my $dds = shift @{$data{ldd}};
    my $i = -1;
    foreach my $key (keys %hashbase) {
        $i++;
        $dds->delete ($key);
        my $insert = $insertions[$i] // next;
        $dds->push ($insert);
        if ($i % 5) {
            $dds->splice ($i);
        }
    }
    [$dds->keys];
}



etc/bench/bench_clone.pl

use Benchmark qw {:all};
use 5.016;
use Data::Dumper;
use Test::More;
use Clone qw //;

use rlib '../../lib';

use List::Unique::DeterministicOrder;
use Clone;


my $nreps = $ARGV[0] || -3;
my $data_size = $ARGV[1] || 1000;
my $run_benchmarks = $ARGV[2] || 1;

#  ratio of insertions to deletions
my $insertion_frac = 0.1;
my $insert_count   = $data_size * $insertion_frac;

srand 1534390472;

my @sorted_keys = ('a' .. 'zzz');
my $base_object = List::Unique::DeterministicOrder->new(data => \@sorted_keys);


my $l1 = lu_clone();
my $l2 = xs_clone();


done_testing();

exit if !$run_benchmarks;


cmpthese (
    $nreps,
    {
        xs_clone  => sub {xs_clone()},
        lu_clone  => sub {lu_clone()},
    }
);

sub lu_clone {
    my $c;
    for my $i (1..1000) {
        $c = $base_object->clone;
    }
    return $c;
}

sub xs_clone {
    my $c;
    for my $i (1..1000) {
        $c = Clone::clone $base_object;
    }
    return $c;
}

lib/Data/Deterministic/Access.pm
package Data::Deterministic::Access;

use 5.010;
use Carp;
use strict;
use warnings;
use List::Util qw /uniq/;

our $VERSION = 0.001;

no autovivification;

sub new {
    my ($package, %args) = @_;

    my $self = {
        hash  => {},
        array => [],
    };
    
    #  use data if we were passed some
    if (my $data = $args{data}) {
        my %hash;
        @hash{@$data} = (0..$#$data);
        #  rebuild the lists if there were dups
        if (scalar keys %hash != scalar @$data) {
            my @uniq = uniq @$data;
            @hash{@uniq} = (0..$#uniq);
            $self->{array} = \@uniq;
        }
        else {
            $self->{array} = [@$data];
        }
        $self->{hash} = \%hash;
    }
    
    return bless $self, $package;
}

sub exists {
    my ($self, $key) = @_;
    return exists $self->{hash}{$key};
}

sub keys {
    my ($self) = @_;
    return wantarray
      ? @{$self->{array}}
      : scalar @{$self->{array}};
}

sub push {
    my ($self, $key) = @_;

    return if exists $self->{hash}{$key};
    
    push @{$self->{array}}, $key;
    $self->{hash}{$key} = $#{$self->{array}};
}

sub pop {
    my ($self) = @_;
    my $key = pop @{$self->{array}};
    delete $self->{hash}{$key};
    return $key;
}

#  returns undef if key not in hash
sub get_key_pos {
    my ($self, $key) = @_;
    return $self->{hash}{$key};
}


#  returns undef if index is out of bounds
sub get_key_at_pos {
    my ($self, $pos) = @_;
    return $self->{array}[$pos];
}

#  does nothing if key does not exist
sub delete {
    my ($self, $key) = @_;
    
    #  get the index while cleaning up
    my $pos = CORE::delete $self->{hash}{$key}
      // return;
    
    my $move_key = CORE::pop @{$self->{array}};
    #  make sure we don't just reinsert the last item
    #  from a single item list
    if ($move_key ne $key) {
        $self->{hash}{$move_key} = $pos;
        $self->{array}[$pos] = $move_key;
    }
    
    return $key;
}

#  Delete the key at the specified position
#  and move the last key into it.
#  Not a true splice, but one day might work
#  on multiple indices.
sub splice {
    my ($self, $pos) = @_;
    
    my $key = $self->{array}[$pos]
      // return;
    
    my $move_key = CORE::pop @{$self->{array}};
    $self->{hash}{$move_key} = $pos;
    $self->{array}[$pos] = $move_key;
    CORE::delete $self->{hash}{$key};
    return $key;
}


sub _paranoia {
    my ($self) = @_;
    
    my $array_len = @{$self->{array}};
    my $hash_len  = CORE::keys %{$self->{hash}};
    croak "array and hash key mismatch" if $array_len != $hash_len;
    
    foreach my $key (@{$self->{array}}) {
        croak "Key mismatch between array and hash lists"
          if !CORE::exists $self->{hash}{$key}; 
    }
    
    return 1;
}

1;

=head1 NAME

Data::Deterministic::Access - Store keys with deterministic order based on insertions and deletions

=head1 VERSION

Version 0.01

=cut

=head1 SYNOPSIS

This module provides a structure to store a set
of keys and be able to access (mostly delete)
them both by key name and by index.

The algorithm used inserts at the end, but
swaps keys around on deletion.  Hence it is
deterministic and repeatable providing the
sequence of insertion and deletions is also
repeated.  

The system does not store any contents with the keys,
so it is more of a "set" structure, but one that
does not do intersections, unions and the like.

So why would one want this in the first place?
In my case it is for a series of random selections
where keys are extracted from a pool of keys,
and sometimes inserted.  The randomisations should
be the same for the same given PRNG sequence.
Binary searches over sorted lists proved very
effective for a while, but bottlenecks started
to manifest when the data sets became
much larger and the number of lists
became both abundant and lengthy.


    use Data::Deterministic::Access;

    my $foo = Data::Deterministic::Access->new(
        data => [qw /foo bar quux fetangle/]
    );
    
    print $foo->keys;
    #  foo bar quux fetangle
    
    $foo->delete ('bar')
    print $foo->keys;
    #  foo fetangle quux 
    
    print $foo->get_key_at_pos(2);
    #  quux
    print $foo->get_key_at_pos(20);
    #  undef
    
    $foo->push ('bardungle')
    print $foo->keys;
    #  foo fetangle quux bardungle

    #  duplicates are stored only once,
    #  just like with a normal hash
    $foo->push ('fetangle')
    print $foo->keys;
    #  foo fetangle quux bardungle
    
    print $foo->exists ('gelert');
    #  false
    
    print $foo->pop;
    #  bardungle
    

=head1 METHODS

Note that most methods take a single argument
(if any), so while the method names look
hash-like, this is purely cosmetic.

=head2 new

Create a new object.
Optionally pass data using the data
keyword.  Duplicate keys are
stored once only.  

=cut

=head2 delete

Deletes the key passed as an argument.

=cut

=head2 exists

True or false for if the key exists.

=cut

=head2 get_key_at_pos

Returns the key at some position.

$foo->get_key_at_pos(5);

=cut

=head2 get_key_pos

Returns the position of a key.

$foo->get_key_pos('quux');

=cut

=head2 keys

Returns the list of keys in list context,
and the number of keys in scalar context.

=cut

=head2 pop

Removes and returns the last key in the list.

=cut

=head2 push

Appends the specified key to the end of the list,
unless it is already in the set.

=cut

=head2 splice

Removes a single key from the set.

=cut


=head1 AUTHOR

Shawn Laffan, C<< <shawnlaffan at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-data-deterministic-access at rt.cpan.org>, or through
the web interface at L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-Deterministic-Access>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Data::Deterministic::Access


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-Deterministic-Access>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Data-Deterministic-Access>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Data-Deterministic-Access>

=item * Search CPAN

L<https://metacpan.org/release/Data-Deterministic-Access>

=back


=head1 ACKNOWLEDGEMENTS

The algorithm used is from
L<https://stackoverflow.com/questions/5682218/data-structure-insert-remove-contains-get-random-element-all-at-o1/5684892#5684892>

=head1 LICENSE AND COPYRIGHT

Copyright 2018 Shawn Laffan 

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut



lib/List/Unique/DeterministicOrder.pm
package List::Unique::DeterministicOrder;

use 5.010;
use Carp;
use strict;
use warnings;
use List::Util 1.45 qw /uniq/;
use Scalar::Util qw /blessed/;

our $VERSION = 0.004;

#no autovivification;

use constant {
    _ARRAY => 0, # ordered keys
    _HASH  => 1, # unordered keys
};

use overload
    q{bool}  => sub { !!%{ $_[0]->[_HASH] } },
    fallback => 1;

sub new {
    my ($package, %args) = @_;

    my $self = [ [], {} ];
    
    #  use data if we were passed some
    if (my $data = $args{data}) {
        my %hash;
        @hash{@$data} = (0..$#$data);
        #  rebuild the lists if there were dups
        if (scalar keys %hash != scalar @$data) {
            my @uniq = uniq @$data;
            @hash{@uniq} = (0..$#uniq);
            $self->[_ARRAY] = \@uniq;
        }
        else {
            $self->[_ARRAY] = [@$data];
        }
        $self->[_HASH] = \%hash;
    }
    
    return bless $self, $package;
}

sub clone {
    my $self = shift;
    my $cloned = bless [], blessed $self;
    $cloned->[_ARRAY] = [@{$self->[_ARRAY]}];
    $cloned->[_HASH]  = {%{$self->[_HASH]}};
    return $cloned;
}

sub exists {
    exists $_[0]->[_HASH]{$_[1]};
}

sub keys {
    wantarray
      ? @{$_[0]->[_ARRAY]}
      : scalar @{$_[0]->[_ARRAY]};
}

sub push {
    return if exists $_[0]->[_HASH]{$_[1]};
    
    push @{$_[0]->[_ARRAY]}, $_[1];
    $_[0]->[_HASH]{$_[1]} = $#{$_[0]->[_ARRAY]};
}

sub pop {
    my $key = pop @{$_[0]->[_ARRAY]};
    delete $_[0]->[_HASH]{$key};
    $key;
}

#  returns undef if key not in hash
sub get_key_pos {
    $_[0]->[_HASH]{$_[1]};
}


#  returns undef if index is out of bounds
sub get_key_at_pos {
    $_[0]->[_ARRAY][$_[1]];
}

#  does nothing if key does not exist
sub delete {
    my ($self, $key) = @_;
    
    #  get the index while cleaning up
    my $pos = CORE::delete $self->[_HASH]{$key}
      // return;

    #  No need to update lists if we are
    #  popping the last index in the array
    return CORE::pop @{$self->[_ARRAY]}
      if $pos == $#{$self->[_ARRAY]};

    $key = CORE::pop @{$self->[_ARRAY]};
    $self->[_HASH]{$key}  = $pos;
    $self->[_ARRAY][$pos] = $key;

    #  obfuscatory but lets us overwrite $key above
    #  and thus avoid an extra scalar variable destruction
    #  (assuming of course that doing so makes a meaningful difference)
    return $_[1];
}

#  Delete the key at the specified position
#  and move the last key into it.
#  Does nothing if key does not exist
sub delete_key_at_pos {
    my ($self, $pos) = @_;
    
    my $key = $self->[_ARRAY][$pos]
      // return;
    
    my $move_key = CORE::pop @{$self->[_ARRAY]};
    CORE::delete $self->[_HASH]{$key};
    
    #  make sure we don't just reinsert the last item
    #  from a single item list
    return $key
      if $pos == @{$self->[_ARRAY]};

    $self->[_HASH]{$move_key} = $pos;
    $self->[_ARRAY][$pos] = $move_key;

    return $key;
}

#  Delete the key at the specified position
#  and move the last key into it.
#  Not a true splice, but one day might work
#  on multiple indices.
#sub splice {
#    my ($self, $pos) = @_;
#    
#    my $key = $self->[_ARRAY][$pos]
#      // return;
#    
#    my $move_key = CORE::pop @{$self->[_ARRAY]};
#    $self->[_HASH]{$move_key} = $pos;
#    $self->[_ARRAY][$pos] = $move_key;
#    CORE::delete $self->[_HASH]{$key};
#    return $key;
#}


sub _paranoia {
    my ($self) = @_;
    
    my $array_len = @{$self->[_ARRAY]};
    my $hash_len  = CORE::keys %{$self->[_HASH]};
    croak "array and hash key mismatch" if $array_len != $hash_len;
    
    foreach my $key (@{$self->[_ARRAY]}) {
        croak "Key mismatch between array and hash lists"
          if !CORE::exists $self->[_HASH]{$key}; 
    }
    
    return 1;
}

1;

=head1 NAME

List::Unique::DeterministicOrder - Store and access 
a list of keys using a deterministic order based on
the sequence of insertions and deletions

=head1 VERSION

Version 0.01

=cut

=head1 SYNOPSIS

This module provides a structure to store a list
of keys, without duplicates, and be able to access
them by either key name or index.


    use List::Unique::DeterministicOrder;

    my $foo = List::Unique::DeterministicOrder->new(
        data => [qw /foo bar quux fetangle/]
    );
    
    print $foo->keys;
    #  foo bar quux fetangle
    
    $foo->delete ('bar')
    print $foo->keys;
    #  foo fetangle quux 
    
    print $foo->get_key_at_pos(2);
    #  quux
    print $foo->get_key_at_pos(20);
    #  undef
    
    $foo->push ('bardungle')
    print $foo->keys;
    #  foo fetangle quux bardungle
    
    #  keys are stored only once,
    #  just like with a normal hash
    $foo->push ('fetangle')
    print $foo->keys;
    #  foo fetangle quux bardungle
    
    print $foo->exists ('gelert');
    #  false
    
    print $foo->pop;
    #  bardungle
    

=head1 DISCUSSION

The algorithm used is from
L<https://stackoverflow.com/questions/5682218/data-structure-insert-remove-contains-get-random-element-all-at-o1/5684892#5684892>

The algorithm used inserts keys at the end, but
swaps keys around on deletion.  Hence it is
deterministic and repeatable, but only if the
sequence of insertions and deletions is replicated
exactly.  

So why would one use this in the first place?
The motivating use-case was a randomisation process
where keys would be selected from a pool of keys,
and sometimes inserted.  e.g. the process might
select and remove the 10th key, then the 257th,
then insert a new key, followed by more selections
and removals.  The randomisations needed to 
produce the same results same for the same given
PRNG sequence for reproducibility purposes.


Using a hash to store the data provides rapid access,
but getting the nth key requires the key list be generated
each time, and Perl's hashes do not provide their
keys in a deterministic
order across all versions and platforms.  
Binary searches over sorted lists proved very
effective for a while, but bottlenecks started
to manifest when the data sets became
much larger and the number of lists
became both abundant and lengthy.

Since the order itself does not matter,
only the ability to replicate it, this module was written.

One could also use L<Hash::Ordered>, but it has the overhead
of storing values, which are not needed here.
I also wrote this module before I benchmarked
against L<Hash::Ordered>.  Regardless, this module is faster
for the example use-case described above - see the
benchmarking results in bench.pl (which is part of
this distribution).  That said, some of the implementation
details have been adapted/borrowed from L<Hash::Ordered>.


=head1 METHODS

Note that most methods take a single argument
(if any), so while the method names look
hash-like, this is essentially cosmetic.
In particular, it does not yet support splicing.  

=head2 new

    $foo->new();
    $foo->new(data => [/a b c d e/]);

Create a new object.
Optionally pass data using the data
keyword.  Duplicate keys are
stored once only.


=cut

=head2 exists

True or false for if the key exists.

=cut

=head2 delete

    $foo->delete('some key');

Deletes the key passed as an argument.
Returns the key name if successful, undef if
the key was not found.

=cut

=head2 delete_key_at_pos

    $foo->delete_key_at_pos(1);

Removes a single key from the set at the specified position.


=cut

=head2 get_key_at_pos

    $foo->get_key_at_pos(5);

Returns the key at some position.

=cut

=head2 get_key_pos

    $foo->get_key_pos('quux');

Returns the position of a key.

=cut

=head2 keys

Returns the list of keys in list context,
and the number of keys in scalar context.

=cut

=head2 pop

    $foo->pop;

Removes and returns the last key in the list.

=cut

=head2 push

    $foo->push('quux');

Appends the specified key to the end of the list,
unless it is already in the list.

=cut

=head2 clone

    $foo->clone;

Creates a clone of the object.

=cut




=head1 AUTHOR

Shawn Laffan, C<< <shawnlaffan at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests via L<https://github.com/shawnlaffan/perl-List-Unique-DeterministicOrder/issues>.  


=head1 ACKNOWLEDGEMENTS

The algorithm used is from
L<https://stackoverflow.com/questions/5682218/data-structure-insert-remove-contains-get-random-element-all-at-o1/5684892#5684892>

Some implementation details have been borrowed/adapted from L<Hash::Ordered>.

=head1 SEE ALSO

L<Hash::Ordered> (and modules listed in its "See Also" section)

L<List::BinarySearch>

L<List::MoreUtils>


=head1 LICENSE AND COPYRIGHT

Copyright 2018 Shawn Laffan 

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

